test/test_nn.py

```diff
@@ -2089,6 +2089,41 @@ def test_embedding_sparse_backward(self):
         self.assertEqual(embedding.weight.grad._indices(), torch.LongTensor([[7, 1, 3, 8, 1, 3]]))
         self.assertEqual(embedding.weight.grad._values(), torch.tensor(1.).expand(6, 3))
 
+    @unittest.skipIf(not TEST_CUDA, "CUDA unavailable")
+    def test_embedding_sparse_half_backward(self):
+        # same as test_embedding_sparse_backward above but testing half types in
+        # cuda. cpu sum not supported.
+        embedding = nn.Embedding(10, 3, sparse=True).half()
```

Comment:
(aside) If a user wanted to use half to save memory, but their table doesn't fit into memory with float tensors, then our API is a little awkward for them (because the weight gets constructed first with float tensors and then gets casted to half...)

Reply:
we should probably add a dtype parameter. I imagine the memory concern is more relevant to cuda though, so at least the user can half().cuda() to shrink before sending to the GPU.

Reply:
That's a good point